File: java/p2/src/p2/lang/plan/DontCare.java
package p2.lang.plan;

import p2.types.basic.Tuple;
import p2.types.function.TupleFunction;

public class DontCare extends Variable {
	public final static String DONTCARE = "_";

	public DontCare(Class type) {
		super(DONTCARE, type);
	}
	
	@Override
	public boolean equals(Object obj) {
		return false;
	}

	@Override
	public TupleFunction function() {
		return new TupleFunction() {
			public Object evaluate(Tuple tuple) {
				return tuple.value(position());
			}

			public Class returnType() {
				return type;
			}
			
		};
	}

}

File: lincoln/ruby/lib/lang/plan/dont_care.rb
require 'lib/lang/plan/variable'
class DontCare < Variable 
  @@DONTCARE = "_"
  @@ids = 0

  def initialize(type, position, loc)
    super("DC" + @@ids.to_s, @type, position, loc);
    @@ids+= 1
  end

  def function()
    throw RuntimeError, "@position < 0" unless @position >= 0
    e_lam = lambda do |t|
      return t.values[@position]
    end

    r_lam = lambda do
      return @type
    end

    tmpClass = Class.new(TupleFunction)
    tmpClass.send :define_method, :evaluate do |tuple|
      e_lam.call(tuple)
    end
    tmpClass.send :define_method, :returnType do 
      r_lam.call
    end
    return tmpClass.new
  end
end